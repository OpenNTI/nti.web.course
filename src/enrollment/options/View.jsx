import './View.scss';
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { decorate } from '@nti/lib-commons';
import { Loading, HOC } from '@nti/web-commons';
import { scoped } from '@nti/lib-locale';

import Store from './Store';
import Administrating from './parts/administrating';
import Enrolled from './parts/enrolled';
import NotEnrolled from './parts/not-enrolled';
import NotAvailable from './parts/not-available';
import Gifting from './parts/gifting';

const t = scoped('course.enrollment.options', {
	unavailable: 'This course is unavailable for enrollment at this time',
	notEnrolledLabel: 'Enroll Today',
});

class CourseEnrollmentOptions extends React.Component {
	static propTypes = {
		className: PropTypes.string,
		catalogEntry: PropTypes.shape({
			getEnrollmentOptions: PropTypes.func,
		}),

		store: PropTypes.shape({
			cleanUp: PropTypes.func,
			setup: PropTypes.func,
			load: PropTypes.func,
		}),
		anonymous: PropTypes.bool,
		loading: PropTypes.bool,
		error: PropTypes.any,
		enrolled: PropTypes.bool,
		administrative: PropTypes.bool,
		options: PropTypes.array,
		access: PropTypes.object,
	};

	componentDidMount() {
		this.setupFor(this.props);
	}

	componentWillUnmount() {
		const { store } = this.props;

		store.cleanUp();
	}

	componentDidUpdate(prevProps) {
		const { catalogEntry: oldEntry } = prevProps;
		const { catalogEntry: newEntry } = this.props;

		if (oldEntry !== newEntry) {
			this.setupFor(this.props);
		}
	}

	setupFor(props) {
		const { catalogEntry, store } = props;

		store.load(catalogEntry);
	}

	onItemChange = () => {
		const { store } = this.props;
		store.setup();
	};

	render() {
		const { className, loading, error, enrolled, catalogEntry } =
			this.props;

		return (
			<HOC.ItemChanges
				item={catalogEntry}
				onItemChanged={this.onItemChange}
			>
				<div
					className={cx('nti-course-enrollment-options', className, {
						'is-enrolled': enrolled,
					})}
				>
					<div className="enrollment-container">
						{loading && <Loading.Spinner />}
						{!loading && error && this.renderError()}
						{!loading && !error && this.renderOptions()}
					</div>
					<div className="gift-container">
						{!loading && !error && this.renderGift()}
					</div>
				</div>
			</HOC.ItemChanges>
		);
	}

	renderError() {
		return <span className="error">{t('unavailable')}</span>;
	}

	renderOptions() {
		const {
			enrolled,
			administrative,
			catalogEntry,
			options,
			access,
			anonymous,
		} = this.props;

		let Cmp = null;

		if (administrative) {
			Cmp = Administrating;
		} else if (!options || !options.length) {
			Cmp = NotAvailable;
		} else if (enrolled) {
			Cmp = Enrolled;
		} else {
			Cmp = NotEnrolled;
		}

		return (
			<Cmp
				anonymous={anonymous}
				options={options}
				catalogEntry={catalogEntry}
				access={access}
			/>
		);
	}

	renderGift() {
		const { options } = this.props;

		if (!options) {
			return null;
		}

		return <Gifting options={options} />;
	}
}

export default decorate(CourseEnrollmentOptions, [
	Store.connect({
		loading: 'loading',
		error: 'error',
		enrolled: 'enrolled',
		administrative: 'administrative',
		anonymous: 'anonymous',
		options: 'options',
		access: 'access',
	}),
]);
