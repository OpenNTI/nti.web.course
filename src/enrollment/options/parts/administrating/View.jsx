import React from 'react';
import PropTypes from 'prop-types';
import {scoped} from '@nti/lib-locale';

import {getCatalogEntryData} from '../../utils';
import Title from '../../common/Title';
import Description from '../../common/Description';

const t = scoped('course.enrollment.options.Administrating', {
	title: 'Administrating',
	description: {
		startDate: 'You are administering this course. Class begins %(fullStartDate)s and will be conducted fully online.',
		noStartDate: 'You are administering this course. Class will be conducted fully online.'
	}
});

export default class CourseEnrollmentOptionsAdministrating extends React.Component {
	static propTypes = {
		catalogEntry: PropTypes.shape({
			getStartDate: PropTypes.func
		})
	}

	render () {
		return (
			<div className="nti-course-enrollment-options-administrating">
				<Title>{t('title')}</Title>
				<Description>{this.renderDescription()}</Description>
			</div>
		);
	}


	renderDescription () {
		const {catalogEntry} = this.props;
		const data = getCatalogEntryData(catalogEntry);

		return data.fullStartDate ?
			t('description.startDate', data) :
			t('description.noStartDate', data);
	}
}
