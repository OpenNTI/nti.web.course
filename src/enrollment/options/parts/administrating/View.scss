.nti-course-enrollment-options-administrating {
	padding-bottom: 1rem;
}
