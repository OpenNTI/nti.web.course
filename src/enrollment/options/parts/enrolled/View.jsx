import React from 'react';
import PropTypes from 'prop-types';

import Current from './Current';
import Upgrade from './Upgrade';

export default class CourseEnrollmentEnrolled extends React.Component {
	static propTypes = {
		options: PropTypes.array,
		cataglogEntry: PropTypes.object
	}

	state = {updating: false}

	componentDidMount () {
		this.setupFor(this.props);
	}

	componentDidUpdate (prevProps) {
		const {options:oldOptions} = prevProps;
		const {options:newOptions} = this.props;

		if (oldOptions !== newOptions) {
			this.setupFor(this.props);
		}
	}


	setupFor (props) {
		const {options} = props;

		this.setState({
			enrolled: options.find(option => option.isEnrolled()),
			upgradeOptions: options.filter(option => option.isAvailable())
		});
	}


	doUpgrade = () => {
		this.setState({
			updating: true
		});
	}


	cancelUpgrade = () => {
		this.setState({
			updating: false
		});
	}


	render () {
		const {updating} = this.state;

		return (
			<div className="nti-coursene-enrollment-options-enrolled">
				{updating ? this.renderUpdate() : this.renderEnrolled()}
			</div>
		);
	}


	renderEnrolled () {
		const {enrolled, upgradeOptions} = this.state;

		if (!enrolled) {
			return null;
		}

		return (
			<Current
				option={enrolled}
				hasUpdates={upgradeOptions && upgradeOptions.length > 1}
				doUpgrade={this.doUpgrade}
			/>
		);
	}


	renderUpdate () {
		const {upgradeOptions} = this.state;

		return (
			<Upgrade
				options={upgradeOptions}
				onCancel={this.cancelUpgrade}
			/>
		);
	}
}
