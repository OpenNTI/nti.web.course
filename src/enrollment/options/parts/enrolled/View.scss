.nti-course-enrollment-options-enrolled {}
