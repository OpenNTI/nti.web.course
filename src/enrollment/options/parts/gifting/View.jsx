import React from 'react';
import PropTypes from 'prop-types';

import Giftable from './Giftable';
import Redeemable from './Redeemable';

export default class CourseEnrollmentOptionsGifting extends React.Component {
	static propTypes = {
		options: PropTypes.arrayOf(
			PropTypes.shape({
				isGiftable: PropTypes.func.isRequired,
				isRedeemable: PropTypes.func.isRequired,
			})
		),
	};

	render() {
		const { options } = this.props;

		const giftable = options.find(option => option.isGiftable());
		const redeemable = options.find(option => option.isRedeemable());

		if (!giftable && !redeemable) {
			return null;
		}

		return (
			<div className="nti-course-enrollment-options-gifting">
				{giftable && <Giftable option={giftable} />}
				{redeemable && <Redeemable option={redeemable} />}
			</div>
		);
	}
}
