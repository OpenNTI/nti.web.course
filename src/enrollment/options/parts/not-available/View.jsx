import React from 'react';

import { scoped } from '@nti/lib-locale';
import { LinkTo } from '@nti/web-routing';
import { useService } from '@nti/web-core';

import Title from '../../common/Title';
import Description from '../../common/Description';
import Button from '../../common/Button';

const t = scoped('course.enrollment.options.NotAvailable', {
	title: 'Invitation only',
	description: 'This course is available via invitation only.',
	redeem: {
		authenticated: 'Redeem Invitation Code',
		unauthenticated: 'Sign In To Get Started',
	},
});

export default function CourseEnrollmentOptionsNotAvailable() {
	const { isAnonymous } = useService();
	const scope = isAnonymous ? 'unauthenticated' : 'authenticated';

	return (
		<div>
			<Title>{t('title')}</Title>
			<Description>{t('description')}</Description>
			<LinkTo.Object object={{ type: 'redeem-course-code' }}>
				<Button>{t(['redeem', scope])}</Button>
			</LinkTo.Object>
		</div>
	);
}
