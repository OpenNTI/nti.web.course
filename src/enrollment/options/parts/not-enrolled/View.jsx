import React from 'react';
import PropTypes from 'prop-types';

import { scoped } from '@nti/lib-locale';

import Title from '../../common/Title';
import OptionList from '../../common/OptionList';

const t = scoped('course.enrollment.options.not-enrolled.View', {
	title: 'Start Learning',
});

export default class CourseEnrollmentOptionsNotEnrolled extends React.Component {
	static propTypes = {
		options: PropTypes.array,
	};

	state = {};

	componentDidMount() {
		this.setupFor(this.props);
	}

	componentDidUpdate(prevProps) {
		const { options: prevOptions } = prevProps;
		const { options: nextOptions } = this.props;

		if (nextOptions !== prevOptions) {
			this.setupFor(this.props);
		}
	}

	setupFor(props) {
		const { options } = this.props;
		const available = options.filter(option => option.isAvailable());

		this.setState({
			override: options.find(option => option.shouldOverride()),
			available: available,
			selected: available[0],
		});
	}

	selectOption = option => {
		this.setState({
			selected: option,
		});
	};

	render() {
		const { override } = this.state;

		return (
			<div className="nti-course-enrollment-options-not-enrolled">
				{override && this.renderOverride()}
				{!override && this.renderOptions()}
			</div>
		);
	}

	renderOptions() {
		const { available, selected } = this.state;

		if (!available) {
			return null;
		}

		const { Description, EnrollButton, SeatLimit } = selected || {};

		return (
			<React.Fragment>
				<Title>{t('title')}</Title>
				<OptionList
					options={available}
					selected={selected}
					selectOption={this.selectOption}
				/>
				{Description && <Description option={selected} />}
				{SeatLimit && <SeatLimit option={selected} />}
				{EnrollButton && <EnrollButton option={selected} />}
			</React.Fragment>
		);
	}

	renderOverride() {}
}
