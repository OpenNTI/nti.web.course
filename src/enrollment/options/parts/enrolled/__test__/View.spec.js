import React from 'react';
import renderer from 'react-test-renderer';

import View from '../View';
import Open from '../../../types/open';
import FiveMinute from '../../../types/five-minute';

const catalogEntry = {
	getStartDate: () => new Date('10/22/2017'),
	getEndDate: () => new Date('10/31/2017')
};

const access = {
	getCreatedTime: () => new Date('10/21/2017')
};

/* eslint-env jest */
describe('Course enrollment enrolled view', () => {
	test('No options', async () => {
		const options = [];

		const cmp = renderer.create(<View catalogEntry={catalogEntry} options={options}/>);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('One option, enrolled, upgrade available', async () => {
		const options = [
			new Open({ enrolled: true, available: true }, access, catalogEntry)
		];

		const cmp = renderer.create(<View catalogEntry={catalogEntry} options={options}/>);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Two options, second enrolled', async () => {
		const options = [
			new Open({ enrolled: true }, access, catalogEntry),
			new FiveMinute({ enrolled: false}, access, catalogEntry)
		];

		const cmp = renderer.create(<View catalogEntry={catalogEntry} options={options}/>);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
