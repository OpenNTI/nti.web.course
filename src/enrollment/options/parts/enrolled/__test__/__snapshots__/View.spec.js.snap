// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course enrollment enrolled view No options 1`] = `
<div
  className="nti-course-enrollment-options-enrolled"
/>
`;

exports[`Course enrollment enrolled view One option, enrolled, upgrade available 1`] = `
<div
  className="nti-course-enrollment-options-enrolled"
>
  <div
    className="nti-course-enrollment-options-enrolled-current"
  >
    <div
      className="nti-course-enrollment-padded-container title"
    >
      <i
        className="icon-check"
      />
      <span>
        Enrolled
      </span>
    </div>
    <div
      className="contents"
    >
      <div
        className="nti-course-enrollment-padded-container nti-course-enrollment-title"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Archived",
          }
        }
      />
      <div
        className="nti-course-enrollment-option-base-enrolled-description"
      >
        <div
          className="nti-course-enrollment-padded-container nti-course-enrollment-description"
          dangerouslySetInnerHTML={
            Object {
              "__html": "The course ended on October 31st, 12:00 AM GMT. The content of this course will remain available for you to review at any time.",
            }
          }
        />
      </div>
    </div>
    <div
      className="nti-link-to-path course-enrollment-options-base-type-drop-button"
      to="#"
    >
      <div
        className="nti-course-enrollment-padded-container course-enrollment-options-action-item"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Remove",
          }
        }
      />
    </div>
    <div
      className="nti-link-to-path"
      to="#"
    >
      <div
        className="nti-course-enrollment-option-button"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Open",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Course enrollment enrolled view Two options, second enrolled 1`] = `
<div
  className="nti-course-enrollment-options-enrolled"
>
  <div
    className="nti-course-enrollment-options-enrolled-current"
  >
    <div
      className="nti-course-enrollment-padded-container title"
    >
      <i
        className="icon-check"
      />
      <span>
        Enrolled
      </span>
    </div>
    <div
      className="contents"
    >
      <div
        className="nti-course-enrollment-padded-container nti-course-enrollment-title"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Archived",
          }
        }
      />
      <div
        className="nti-course-enrollment-option-base-enrolled-description"
      >
        <div
          className="nti-course-enrollment-padded-container nti-course-enrollment-description"
          dangerouslySetInnerHTML={
            Object {
              "__html": "The course ended on October 31st, 12:00 AM GMT. The content of this course will remain available for you to review at any time.",
            }
          }
        />
      </div>
    </div>
    <div
      className="nti-link-to-path course-enrollment-options-base-type-drop-button"
      to="#"
    >
      <div
        className="nti-course-enrollment-padded-container course-enrollment-options-action-item"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Remove",
          }
        }
      />
    </div>
    <div
      className="nti-link-to-path"
      to="#"
    >
      <div
        className="nti-course-enrollment-option-button"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Open",
          }
        }
      />
    </div>
  </div>
</div>
`;
